c++/src/kj/compat/brotli.c++

```diff
@@ -0,0 +1,355 @@
+    }
+    BrotliDecoderResult result =
+        BrotliDecoderDecompressStream(dctx, &available_in, &next_in, &available_out, &next_out,
+                                      nullptr);
+    if (result == BROTLI_DECODER_RESULT_ERROR) {
+        // Note: Unlike BrotliInputStream, this will implicitly reject trailing data during
+        // decompression, matching the behavior for gzip.
+        KJ_FAIL_REQUIRE("brotli decompression failed",
+                        BrotliDecoderErrorString(BrotliDecoderGetErrorCode(dctx)));
+    }
+  } else {
+    BROTLI_BOOL result = BrotliEncoderCompressStream(cctx, flush, &available_in, &next_in,
+                                                     &available_out, &next_out, nullptr);
+    KJ_REQUIRE(result == BROTLI_TRUE, "brotli compression failed");
+  }
+
+  // The 'ok' parameter represented by the first parameter of the tuple indicates that pumpOnce()
+  // should be called again as more output data can be produced. This is the case when the stream
+  // is not finished and there is either pending output data (that didn't fit into the buffer) or
+  // input that has not been processed yet.
+
+  return kj::tuple(
+      (compressing && BrotliEncoderHasMoreOutput(cctx)) ||
+      (!compressing && BrotliDecoderHasMoreOutput(dctx)),
+       kj::arrayPtr(buffer, sizeof(buffer) - available_out));
+}
+
+}  // namespace _ (private)
+
+// =======================================================================================
+
+BrotliInputStream::BrotliInputStream(InputStream& inner, kj::Maybe<int> _windowBits)
+    : inner(inner), next_in(nullptr), available_in(0) {
+  windowBits = _windowBits.orDefault(_::KJ_BROTLI_MAX_DEC_WBITS);
+  KJ_REQUIRE(windowBits >= BROTLI_MIN_WINDOW_BITS && windowBits <= BROTLI_MAX_WINDOW_BITS,
+      "invalid brotli window size (window bits", windowBits, ")");
+  ctx = BrotliDecoderCreateInstance(nullptr, nullptr, nullptr);
+  KJ_REQUIRE(ctx, "brotli state allocation failed");
+}
+
+BrotliInputStream::~BrotliInputStream() noexcept(false) {
+  BrotliDecoderDestroyInstance(ctx);
+}
+
+size_t BrotliInputStream::tryRead(void* out, size_t minBytes, size_t maxBytes) {
+  if (maxBytes == 0) return size_t(0);
+
+  return readImpl(reinterpret_cast<byte*>(out), minBytes, maxBytes, 0);
+}
+
+size_t BrotliInputStream::readImpl(
+    byte* out, size_t minBytes, size_t maxBytes, size_t alreadyRead) {
+  // Ask for more input unless there is pending output
+  if (available_in == 0 && !BrotliDecoderHasMoreOutput(ctx)) {
+    size_t amount = inner.tryRead(buffer, 1, sizeof(buffer));
+    if (amount == 0) {
+      KJ_REQUIRE(atValidEndpoint, "brotli compressed stream ended prematurely");
+      return alreadyRead;
+    } else {
+      next_in = buffer;
+      available_in = amount;
+    }
+  }
+
+  byte* next_out = out;
+  size_t available_out = maxBytes;
+  // Check window bits
+  if (firstInput) {
+    firstInput = false;
+    int streamWbits = _::get_brotli_window_bits(next_in[0]);
```

Comment:
I added `availableIn > 0` to the if condition in all spots where the window bits are checked – this ensures that if `pumpOnce()` is called before an input is set it will effectively perform a no-op and we only do the window bits check if data is available.